File: python/protein-translation/protein_translation.py
from itertools import takewhile
from textwrap import wrap


catalogue = {
    'AUG': 'Methionine',
    'UUC': 'Phenylalanine',
    'UUU': 'Phenylalanine',
    'UUA': 'Leucine',
    'UUG': 'Leucine',
    'UCU': 'Serine',
    'UCC': 'Serine',
    'UCA': 'Serine',
    'UCG': 'Serine',
    'UAC': 'Tyrosine',
    'UAU': 'Tyrosine',
    'UGC': 'Cysteine',
    'UGU': 'Cysteine',
    'UGG': 'Tryptophan',
}


def is_not_stop(pattern):
    return pattern not in ('UAG', 'UAA', 'UGA')


def proteins(strand):
    print(wrap(strand, 3))
    return [protein
            for codon in takewhile(is_not_stop, wrap(strand, 3)) # considering codons only while a 'STOP' codon is not reached
            for protein in (catalogue.get(codon),) # tuple with a single element
            if protein] # eliminating codons not existing in catalogue

File: javascript/protein-translation/protein-translation.js
const codonProteinMap = new Map([
  ['AUG', 'Methionine'],
  ['UUU', 'Phenylalanine'],
  ['UUC', 'Phenylalanine'],
  ['UUA', 'Leucine'],
  ['UUG', 'Leucine'],
  ['UCU', 'Serine'],
  ['UCC', 'Serine'],
  ['UCA', 'Serine'],
  ['UCG', 'Serine'],
  ['UAU', 'Tyrosine'],
  ['UAC', 'Tyrosine'],
  ['UGU', 'Cysteine'],
  ['UGC', 'Cysteine'],
  ['UGG', 'Tryptophan'],
  ['UAA', 'STOP'],
  ['UAG', 'STOP'],
  ['UGA', 'STOP']
]);

export const translate = (sentence='') => {
  if (sentence.length % 3 !== 0) throw new Error('Invalid codon');
  let proteins = [];
  for (let i = 0; i < sentence.length; i += 3) {
    const codon = sentence.slice(i, i + 3);
    const protein = codonProteinMap.get(codon);
    if (protein === 'STOP') break;
    else if (protein) proteins.push(protein);
    else throw new Error('Invalid codon');
  }
  return proteins;
};
